From: Herbert Xu <herbert.xu@redhat.com>
Date: Thu, 26 May 2011 03:23:23 -0400
Subject: [net] core: Fix memory leak/corruption on VLAN GRO_DROP
Message-id: <20110526032322.GA14445@gondor.apana.org.au>
Patchwork-id: 4582
O-Subject: [kernel team] [RHEL5 PATCH] net: Fix memory leak/corruption on VLAN GRO_DROP
Bugzilla: 691565
CVE: CVE-2011-1576
RH-Acked-by: Thomas Graf <tgraf@redhat.com>
RH-Acked-by: Andy Gospodarek <gospo@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

RHEL5 bugzilla #691565

This patch fixes a serious memory leak/corruption bug in RHEL5.
This bug used to exist upstream as well, but was fixed when the
VLAN code path was restructured in 2010.

The approach here is different from upstream in order to minimise
the effect on the code.

net: Fix memory leak/corruption on VLAN GRO_DROP

The function napi_reuse_skb is only meant to be used for packets
merged by GRO.  Using it on the VLAN path will lead to memory
leaks/corruption.  This patch is based on Jay Vosburgh's patch,
and it fixes the problem by calling kfree_skb on the VLAN GRO_DROP
path instead of napi_reuse_skb.

Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>

Cheers,

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/core/dev.c b/net/core/dev.c
index 0be4182..1d3e58e 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2260,7 +2260,8 @@ int napi_frags_finish(struct napi_struct *napi, struct sk_buff *skb, int ret)
 
 	case GRO_DROP:
 		err = NET_RX_DROP;
-		/* fall through */
+		kfree_skb(skb);
+		break;
 
 	case GRO_MERGED_FREE:
 		napi_reuse_skb(napi, skb);
